### PR TITLE
Remove fetch after upload

### LINKED ISSUE #1971: Consider avoiding re-fetch of media after upload
Consider avoiding the media fetch after upload, and instead retrieve the updated resource info from the response of the upload api call.

https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js#L124

This solves two problems:

1. As a part of the larger epic to avoid re-renders, prevent the 'uploadMedia' function reference from being re-created when pagingNum changes.
2. fetch after upload uses the current 'pagingNum', which doesn't update the latest media if the gallery is scrolled down the page.

See also: issue https://github.com/google/web-stories-wp/issues/1971 and PR https://github.com/google/web-stories-wp/pull/1848.

A corner case to consider:
1. User search for cats
2. User filter media library by video.
3. User uploads an image called dog.jpg.

Should the uploaded image be added to the top of the list (and the user scrolled to the top), or should the item disappear? Or should the search be cleared?

### PR DESCRIPTION
**Pull request by cvolzke4**
## Summary

Remove fetch after upload. Instead, retrieve the updated resource info from the response of the upload api call.

This solves two problems:
* As a part of the larger epic to avoid re-renders, prevent the 'uploadMedia' function reference from being re-created when pagingNum changes.
* fetch after upload uses the current 'pagingNum', which doesn't update the latest media if the gallery is scrolled down the page.

### CODE REVIEW

**Comment by spacedmonkey** [comment]
The fetch after upload was added for two reasons. 

- Why video poster is generated and uploaded, we needed up force the media library to reload. But I believe this has since been fixed. 
- If images have been uploaded in the WP media dialog, we need to force this to reload. 

@cvolzke4  you need to take into account this case in this PR, I believe.

**Comment by cvolzke4** [comment]
> The fetch after upload was added for two reasons.
> 
> * Why video poster is generated and uploaded, we needed up force the media library to reload. But I believe this has since been fixed.
> * If images have been uploaded in the WP media dialog, we need to force this to reload.
> 
> @cvolzke4 you need to take into account this case in this PR, I believe.

In these changes, the return value from 'await uploadFile(localFile.file)' is used to update the media directly. I tested this by drag/drop of an image into the media pane, and checked that the attributes are consistent with a server reload.

When using the 'Upload' button to upload into WP, the media pane is reloaded elsewhere, but it's still reloaded as expected.

**Comment by spacedmonkey** [comment]
Screenshot 2020-05-21 at 15 41 22
Whenever I try and upload an image via drag and drop or pasting, I get this error.

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js [hunk 1]
```diff
@@ -104,24 +104,37 @@ function useUploadMedia({ media, pagingNum, mediaType, fetchMedia, setMedia }) {
...
             onUploadedFile({
-              resource: getResourceFromAttachment(fileUploaded),
+              resource: fileUploaded,
               element,
             });
           });
         }
-        fetchMedia({ pagingNum, mediaType }, setMedia);
+
+        const uploadedFilesMap = Map(
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js]:
```suggestion
        const uploadedFilesMap = new Map(
```

Comment (spacedmonkey) [reply assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js]: Screenshot 2020-05-21 at 15 41 22
Fixes this issue, I was seeing.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Even with the `new Map` fix, this PR didnt work for me. The progress bar on the upload remained forever and was never replaced with uploaded version.

**Comment by cvolzke4** [comment]
> Screenshot 2020-05-21 at 15 41 22
> Whenever I try and upload an image via drag and drop or pasting, I get this error.

My bad, I thought I was working in dart again :)
Fixed.

**Comment by spacedmonkey** [comment]
How does this PR work when a filter is active? 
Case.
1. User search for cats
1. User filter media library by video. 
1. User uploads a image called. dog.jpg.

Currently this just add the image to the top of the list, is that correct? I am not sure how we work around this, unless if does the refetch if a filter or search is set overwise, just adds to the top?

**Comment by spacedmonkey** [comment]
Screenshot 2020-05-21 at 15 52 52
Still seeing the progress bar forever more.

**Comment by cvolzke4** [comment]
> Even with the `new Map` fix, this PR didnt work for me. The progress bar on the upload remained forever and was never replaced with uploaded version.

I was too fast to reply. This has been fixed, please try again.

**Comment by spacedmonkey** [comment]
Also worthing noting that seems like force pushing here to rewrite commits. Please avoid doing this, as making pulling your change hard as I have to delete the branch and fetch it again.

**Comment by cvolzke4** [comment]
> How does this PR work when a filter is active?
> Case.
> 
> 1. User search for cats
> 2. User filter media library by video.
> 3. User uploads a image called. dog.jpg.
> 
> Currently this just add the image to the top of the list, is that correct? I am not sure how we work around this, unless if does the refetch if a filter or search is set overwise, just adds to the top?

It's a good question, this may be a case for re-fetching (or re-apply the search in memory, if that's easy to do). Another way to look at it is as a 'feature', so that the user can see what they just uploaded. It may appear to be a glitch if an upload finishes, and then the image (or images) suddenly disappear.

When a search is active inside the WP uploader, the image isn't displayed, but image details appear on a pane to the right of the list (the last uploaded image if multiple were dropped).

It may not matter very much, as the user can re-enter the search term to re-apply the filter if needed.

This sounds like a question for UX, WDYT?

**Comment by cvolzke4** [comment]
> Also worthing noting that seems like force pushing here to rewrite commits. Please avoid doing this, as making pulling your change hard as I have to delete the branch and fetch it again.

Thanks for letting me know. I was under the impression that this only affects a pull if another contributor is making changes to the same branch.

**Comment by cvolzke4** [comment]
> Still seeing the progress bar forever more

I don't think this is the updated version, try refreshing the browser.

**Comment by spacedmonkey** [comment]
Changes like this really need a ticket first and a the solution to be discussed before any code is written. If that was the case I would have pointed out the pitfalls and why we had to be the refetch in the first place. 

1. Can we create a ticket so this work can be tracked. 
2. Confirm UX after upload with Sam. 
3. Write a set of tests, that proves this doesn't break anything. A change this wide reaching will likely have breakages. 

We need to validate that this doesn't break the 5 kinds of upload
- Upload via drag into media library. 
- Upload via drag into canvas. 
- Upload via paste ( only works on Mac OS )
- Upload via WP media modal. 
- Upload via poster generation.

**Comment by cvolzke4** [comment]
> Changes like this really need a ticket first and a the solution to be discussed before any code is written. If that was the case I would have pointed out the pitfalls and why we had to be the refetch in the first place.
> 
> 1. Can we create a ticket so this work can be tracked.
> 2. Confirm UX after upload with Sam.
> 3. Write a set of tests, that proves this doesn't break anything. A change this wide reaching will likely have breakages.
> 
> We need to validate that this doesn't break the 5 kinds of upload
> 
> * Upload via drag into media library.
> * Upload via drag into canvas.
> * Upload via paste ( only works on Mac OS )
> * Upload via WP media modal.
> * Upload via poster generation.

Thanks Jonny.

This change relates to the performance ticket (which I've linked now, I should have done this earlier). I'll raise another ticket Monday to address the specific change. I had started this PR as a way to communicate this in a comment elsewhere, and then finished the implementation since it wasn't very much work and I could receive feedback on the approach, which you have given (thank you). I'll write some new tests once the behavior has been confirmed with Sam.

Please bear with me while we find our feet working with github, linking issues, getting feedback etc, as you can understand it's a very different environment to what we are used to.

I've just tested the upload types that you mentioned, including upload of multiple files.

There's a few other issues that I noticed while testing:
* fetchMedia depends on pagingNum, which causes re-renders during scrolling (the issue I'm trying to fix).
* Uploading a video doesn't focus the 'Video' tab, but it probably should.
* fetchMedia uses pagingNum to load the last page of results, and prepends them to the list of media. This change would also fix that.
* Pasting an mp4 video into the canvas pastes the filename, but not the content.
* Uploading a video into the canvas froze chrome (appears to be a once off, I can no longer reproduce it).

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js [hunk 2]
```diff
@@ -104,24 +104,37 @@ function useUploadMedia({ media, pagingNum, mediaType, fetchMedia, setMedia }) {
...
         }
-        fetchMedia({ pagingNum, mediaType }, setMedia);
+
+        const uploadedFilesMap = new Map(
+          uploadedFiles.map(({ localResource, fileUploaded }) => [
+            localResource,
+            fileUploaded,
+          ])
+        );
+        setMedia({
```
Comment (dvoytenko) [inline assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js:133]: Seems like reshuffling the whole array of media elements is a lot. I think we already have `updateMediaElement`. Can we use it here?

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js [hunk 3]
```diff
@@ -87,12 +88,11 @@ function useUploadMedia({ media, pagingNum, mediaType, fetchMedia, setMedia }) {
             return { localResource, file, element };
           });
         }
-        setMedia({
-          media: [
-            ...localFiles.map(({ localResource }) => localResource),
-            ...media,
-          ],
-        });
+        updatedMedia = [
```
Comment (dvoytenko) [inline assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js:91]: Would it make sense to have instead a targeted method `prependMedia` and resolve array logic and stuff on the reducer level?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

**Comment by spacedmonkey** [comment]
Honestly, there are still a lot of question regarding the UX of the media library about expected behavior. Like

What you are scrolled down say 200th item in the media list, does the preview appear at the top of the list? 
Should a preview should for a video when image has been filtered. 
What the preview upload appear at all if a search is active? 
Are the any cases that where the filters and search need to be reset?

I don't think that we can work with the idea that current UX is correct as most of it was written before any of the current designs were implemented. 

Until UX is confirmed, is hard to know if this PR is doing that correct behavior.

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js (hunk 3 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js:91]: That would be a better way, this is an existing issue that can be addressed in a separate PR. I've raised an issue: https://github.com/google/web-stories-wp/issues/2072

**Comment by dvoytenko** [comment]
@cvolzke4 trying to catch up on this. is this ready to be rereviewed? I'm generally LG with changes, but there's a longish threads with other reviewers...

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js (hunk 2 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js:133]: As it is today, the reducer for updateMediaElement is also O(n) for each media element. But we should be able to refactor it to take more than one media element at a time.

Here is an issue to address this: https://github.com/google/web-stories-wp/issues/2072

**Comment by cvolzke4** [comment]
> @cvolzke4 trying to catch up on this. is this ready to be rereviewed? I'm generally LG with changes, but there's a longish threads with other reviewers...

Thanks Dima, I was going to ping you about this today :)
UX has approved this approach (see https://github.com/google/web-stories-wp/issues/1971), so this PR is ready for review.

**Review by dvoytenko — APPROVED** [review]

**Review by spacedmonkey — APPROVED** [review]
LGTM

**Pull request merged**